lib/archethic/contracts/interpreter/library/common/http_impl.ex

```diff
@@ -0,0 +1,156 @@
+defmodule Archethic.Contracts.Interpreter.Library.Common.HttpImpl do
+  @moduledoc """
+  Http client for the Smart Contracts.
+  Implements AEIP-20.
+
+  Mint library is processless so in order to not mess with
+  other processes, we use it from inside a Task.
+  """
+
+  alias Archethic.Contracts.Interpreter.Library
+  alias Archethic.Contracts.Interpreter.Library.Common.Http
+  alias Archethic.TaskSupervisor
+
+  @behaviour Http
+  @threshold 256 * 1024
+
+  @impl Http
+  def fetch(uri) do
+    task =
+      Task.Supervisor.async_nolink(
+        TaskSupervisor,
+        fn -> do_fetch(uri) end
+      )
+
+    case Task.yield(task, 2_000) || Task.shutdown(task) do
+      {:ok, {:ok, reply}} ->
+        reply
+
+      {:ok, {:error, :threshold_reached}} ->
+        raise Library.Error, message: "Http.fetch/1 response is bigger than threshold"
+
+      {:ok, {:error, _}} ->
+        # Mint.HTTP.connect error
+        # Mint.HTTP.stream error
+        raise Library.Error, message: "Http.fetch/1 failed"
+
+      {:ok, {:error, _, _}} ->
+        # Mint.HTTP.request error
+        raise Library.Error, message: "Http.fetch/1 failed"
+
+      nil ->
+        # Task.shutdown
+        raise Library.Error, message: "Http.fetch/1 timed out for url: #{uri}"
+    end
+  end
+
+  @impl Http
+  def fetch_many(uris) do
+    uris_count = length(uris)
+
+    if uris_count > 5 do
+      raise Library.Error, message: "Http.fetch_many/1 was called with too many urls"
+    else
+      Task.Supervisor.async_stream_nolink(
+        TaskSupervisor,
+        uris,
+        &fetch/1,
+        ordered: true,
+        max_concurrency: 5
+      )
+      |> Enum.to_list()
+      # count the number of bytes to be able to send a error too large
+      # this is sub optimal because miners might still download threshold N times before returning the error
+      # TODO: improve this
+      |> Enum.reduce({0, []}, fn
+        {:exit, {e, _stacktrace}}, _ ->
+          # if any fetch/1 raised, we raise
+          raise Library.Error, message: e.message
+
+        {:ok, map}, {bytes_acc, result_acc} ->
+          bytes =
+            case map["body"] do
+              nil -> 0
+              body -> byte_size(body)
+            end
+
+          {bytes_acc + bytes, result_acc ++ [map]}
+      end)
+      |> then(fn {bytes_total, result} ->
+        if bytes_total > @threshold do
+          raise Library.Error,
+            message: "Http.fetch_many/1 sum of responses is bigger than threshold"
+        else
+          result
+        end
+      end)
+    end
+  end
+
+  defp do_fetch(url) do
+    uri = URI.parse(url)
+
+    # we use the transport_opts to be able to test (MIX_ENV=test) with self signed certificates
+    conn_opts = [
+      transport_opts:
+        Application.get_env(:archethic, __MODULE__, [])
+        |> Keyword.get(:transport_opts, [])
+    ]
+
+    with :ok <- check_scheme(uri.scheme),
```

Comment:
Maybe you could rename this function to `validate_scheme` as we often use this wording for function returning `:ok | {:error, reason}`